Before you turn this homework in, make sure everything runs as expected. First, **restart the kernel** and then **run all cells** (in the menubar, select Kernel$\rightarrow$Restart \& Run all). If you get error messages, the grader will get the same errors and won't be able to grade your work. 

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE". You may delete the line(s) that say "**raise** NotImplementedError()" and replace them with your actual work. Write all your work in the cells provided in this notebook. **DO NOT** add or delete any cells. 

Please fill in your name in the cell below.

In [1]:
NAME = "Zander Bonnet"

---

# Homework 4

In all problems below, please comment your code sufficiently well so that the grader can follow what you are doing with ease. For non-coding answers, please make sure to formulate your explanation and answers in the form of complete English sentences. It is not sufficient to just leave comments in the code (without a full sentence explanations in problems that ask for those) or vice versa (to leave full sentence explanations but not code comments). You will need both to get full credit. 

## Problem 1

Write a function named "border" that takes as its input two integers n and m ($1 \le n,m \le 100$) and outputs a NumPy array of size $(n,m)$. Your array should be filled with zeros except for the "border" (that is, the first and last column and the first and last row) which should be filled with ones. 

In [2]:
import numpy as np

def border(n, m):
    """Creates an array with shape (n, m) that is all zeros
    except for the border (i.e., the first and last rows and
    columns), which should be filled with ones."""
    # YOUR CODE HERE
    array = np.zeros((n,m)) #Makes an array of size nxm of all zeros
    for row in range(n): #Iterates over all rows
        for col in range(m): #Iterates over all colums in the row
            if((row == 0) | (row == n - 1) | (col == 0) | (col == m-1) ): #If an end case
                array[row,col] = 1       #Set to 1
    return(array)

In [3]:
# check that your function works as expected (if it does this check should return "True")
np.array_equal(border(3,3), np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]]))


True

## Problem 2

(a) (2 points) Suppose that x and y are both NumPy arrays of size $(1,n)$. 
Write a function that computes Pearson's sample correlation coefficient $r$ 
between the entries in x and y. 

Recall, that Pearson's sample correlation coefficient is defined as 
$$r(x,y) = \frac{\sum_{i=1}^n (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum\limits_{i=1}^n (x_i - \bar{x})^2 \sum\limits_{i=1}^n(y_i - \bar{y})^2}} $$

**Note:** Do *not* use the built in Numpy function ```np.corrcoef()```. Instead, please use Numpy ufuncs for sums, means, and squareroots.

In [4]:
def pearsons_correlation(x, y):
    """Computes Pearson's correlation coefficient between vectors x and y."""
    # YOUR CODE HERE
    meanX = np.mean(x) #Gets mean of X
    meanY = np.mean(y) #Gets mean of Y
    top = 0    #Initilize variables
    botX = 0
    botY = 0
    for i in range(np.size(x,1)): #For all values of the columns
        top += ((x[0,i] - meanX) * (y[0,i] - meanY)) #Finds top of equation
        botX += (x[0,i] - meanX)**2 #Finds bottom of the x's
        botY += (y[0,i]-meanY)**2 #Finds bottom of the y's
    bot = (botX * botY)**.5 #Sqrt of the product of variances
    return top/bot
x = np.random.random((1,10))
y = np.random.random((1,10)) 
print(np.corrcoef(x,y)[0,1])
print(pearsons_correlation(x,y))

0.5540150762232449
0.5540150762232449


In [5]:
# check that your function works
# if it does, the comparison below should return "True"

import numpy as np

np.random.seed(10)
n = 10
x = np.random.random((1,10))
y = np.random.random((1,10))    

round(pearsons_correlation(x,y),6) == round(np.corrcoef(x,y)[0,1],6)
# we're rounding, because the built-in NumPy method for correlation 
# may use more efficient arithmetic than we do in our home-made function. 

True

(b) (6 points) Create a NumPy array filled with uniform(0,1) random values with n rows and m columns (for the combinations of n and m listed below). In each case, find the maximum of the Pearson correlation coefficients for each of the $\binom{m}{2}$ pairs of columns in your array. 

That is, if the columns of your array are $x_1, \ldots, x_m$, find 
$$\max(r) = \max\limits_{a,b \in \{1,\ldots,m\}, a \neq b} r(x_a, x_b)$$

Enter the values you find into the table below (rounded to two after decimal digits).

In [6]:
import numpy as np
np.random.seed(10)

def max_corr(n,m):
    array = np.random.random((n,m))
    """find maximum correlation between columns of (n,m) random uniform (0,1) array"""
    # YOUR CODE HERE
    maximum = -100000 # Sets max value to something that will be overwritten
    for col in range(np.size(array,1)-1): #For the first col
        for col2 in range(col+1,np.size(array,1)): #For every col after the first col
            cor = np.corrcoef(array[:,col],array[:,col2])[0,1] #Finds the corelation
            if(cor > maximum): #If the max set it to max
                maximum = cor    
    return maximum

print(max_corr(10,20))
print(max_corr(10,50))
print(max_corr(10,100))
print(max_corr(100,20))
print(max_corr(100,50))
print(max_corr(100,100))

0.7339282571155443
0.8176873855394015
0.9272644039978727
0.26424082642684377
0.35895309673824183
0.3861266885311634


|  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  | |
| 10  | 50  |  |
| 10  | 100  |  |
| 100  | 20  |  |
| 100  | 50  |  |
| 100  | 100  |  |

Copy and paste this table to the Markdown cell below and complete it. Describe what you observe. Is the maximal correlation between columns (representing features here) an increasing function of n (number of rows or cases)? Is an increasing function of m (number of features)?

Since the data are entirely random, please note that any correlations you observe are entirely [spurious](https://www.tylervigen.com/spurious-correlations). 

|  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  |0.7339282571155443 |
| 10  | 50  | 0.8176873855394015 |
| 10  | 100  | 0.9272644039978727 |
| 100  | 20  | 0.26424082642684377 |
| 100  | 50  | 0.35895309673824183 |
| 100  | 100  | 0.3861266885311634 |

I observed as "m" increased the max increased and as "n" increased the max decreased.


## Problem 3

(4 points) Suppose that $X$ is a matrix with n rows and m columns. In statistics, these matrices arise freqently when we collect (numerical) data on m variables which are each observed on n independent individuals. Suppose we denote the columns of $X = (x_1, \ldots, x_m)$. Then the covariance matrix of $X$ is defined as 

$$ \mbox{Cov}(X) = \left( \begin{array}{cccc}
Var(x_1) & Cov(x_1, x_2) & \cdots & Cov(x_1, x_m) \\
Cov(x_2, x_1) & Var(x_2) & \cdots & Cov(x_2, x_m) \\
\vdots & \vdots & \ddots & \vdots \\
Cov(x_m, x_1) & Cov(x_m, x_2) & \cdots & Var(x_m) 
\end{array} \right) $$

where Var($x_i$) is the sample variance of the entries in the $i^{th}$ column of $X$ and Cov($x_i, x_j$) ($i \neq j$) is the sample covariance of the entries in columns $i$ and $j$. 

$$ Cov(x_i, x_j) = \frac{1}{n-1}\sum\limits_{k=1}^n (x_{ki} - \bar{x}_i)(x_{kj}-\bar{x_j})$$

That is, the covariance matrix is a square $m\times m$ matrix whose diagonal entries are the sample variances of the columns of $X$ and whose off-diagonal entries are covariances between two columns of $X$, respectively. 

Write *your own* function that takes a Numpy array of shape (n,m) as its input and returns the  Covariance matrix of the array. Do not use any built-in functions to compute variance and/or covariance. Write your own functions, instead. It's ok to use NumPy routines for sums, squares, or square-roots. 

Check your work by comparing your result to that of the ```np.cov()``` function applied to the same input.

In [7]:
def covariance_matrix(X):
    """ Finds the covariance matrix of (n,m) shaped array X"""
    # YOUR CODE HERE
    array = np.zeros((np.size(X,1),np.size(X,1))) #Makes a mxm matrix of all zeros
    for col in range(np.size(X,1)): #For the first column to be looked at 
        meanCol = np.mean(X[:,col]) #Finds the mean of the first locked column
        for col2 in range(np.size(X,1)): #For the second column
            su = 0 #Sets the sum to zero
            meanCol2 = np.mean(X[:,col2]) #Gets the mean for the second locked column
            for i in range(np.size(X,0)): #For every value in the columns
                su += (X[i,col] - meanCol) * (X[i,col2] - meanCol2) #The formula
            array[col,col2] = su/(np.size(X,0)-1) #Gets the var/cov and puts in the array
    return array

In [8]:
## check your work
X = np.random.random((5,3))  # or make up some other matrix for X 
np.array_equal(np.round(np.cov(X.T),4), np.round(covariance_matrix(X),4)) 

# again, we're rounding because your "by-hand" computation will differ slightly from NumPy's internal computations

True

## Problem 4

The code below generates a NumPy array X of shape (12,8) filled with integers. Please don't change the seed.

In [9]:
import numpy as np

np.random.seed(10)
X = np.random.randint(10,size = (12,8)) 
X

array([[9, 4, 0, 1, 9, 0, 1, 8],
       [9, 0, 8, 6, 4, 3, 0, 4],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]])

(a) (2 points) Write a function called "swap" that takes as its input the array $X$ and two integers n and m ($ 1 \le n,m \le 12, n \neq m$) and returns the matrix X in which rows n and m have been swapped. 

In [10]:
def swap(X,n,m):
    """swaps rows n and m of array X"""
    # YOUR CODE HERE
    swap1 = X[n-1,:].copy() #Makes a copy of the first to be swapped
    swap2 = X[m-1,:].copy() #Makes a copy of teh second to be swaped
    for i in range(np.size(X,1)): #Swaps all the elements one by one
        X[n-1,i] = swap2[i]
        X[m-1,i] = swap1[i]
    return X

In [11]:
# check your work
np.array_equal(swap(X,1,2), np.array([[9, 0, 8, 6, 4, 3, 0, 4],
       [9, 4, 0, 1, 9, 0, 1, 8],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]]))  


True

(b) (2 points) Write a function called "sort_by_column" that takes as its input the array X and an integer k ($1\le k \le 8$) and sorts the rows of X by the values of column k. Numbers that were in the same row before, should still be in the same row. Don't worry about ties - if there are ties, I don't care how you order them. 

In [12]:
def sort_by_column(X,k):
    """ sort rows of X by values in column k"""
    # YOUR CODE HERE
    raise NotImplementedError()